File: app/src/main/java/com/patientapp/health/data/UserRepository.java
package com.patientapp.health.data;

import androidx.lifecycle.MutableLiveData;

import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.ListenerRegistration;

import java.util.ArrayList;
import java.util.List;

public class UserRepository {

    private final FirebaseFirestore firestore;

    public UserRepository() {
        this.firestore = FirebaseFirestore.getInstance();
    }

    public ListenerRegistration observePatientsByDoctor(String doctorId,
                                                        MutableLiveData<List<User>> liveData) {
        return firestore.collection(FirestoreConstants.USERS)
                .whereEqualTo(FirestoreConstants.ROLE, UserRole.PATIENT.name())
                .whereEqualTo(FirestoreConstants.DOCTOR_ID, doctorId)
                .addSnapshotListener((snapshot, error) -> {
                    if (error != null || snapshot == null) {
                        liveData.setValue(new ArrayList<>());
                        return;
                    }
                    List<User> users = new ArrayList<>();
                    for (DocumentSnapshot doc : snapshot.getDocuments()) {
                        User user = DocumentParser.toUser(doc);
                        if (user != null) users.add(user);
                    }
                    liveData.setValue(users);
                });
    }

    public void updateFcmToken(String userId, String token) {
        firestore.collection(FirestoreConstants.USERS).document(userId)
                .update(FirestoreConstants.FCM_TOKEN, token);
    }
}

File: app/src/main/java/com/patientapp/health/data/UserRepository.kt
package com.patientapp.health.data

import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.tasks.await

class UserRepository {

    private val firestore = FirebaseFirestore.getInstance()

    fun getPatientsByDoctor(doctorId: String): Flow<List<User>> = callbackFlow {
        val listener = firestore.collection(FirestoreConstants.USERS)
            .whereEqualTo(FirestoreConstants.ROLE, UserRole.PATIENT.name)
            .whereEqualTo(FirestoreConstants.DOCTOR_ID, doctorId)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    trySend(emptyList())
                    return@addSnapshotListener
                }
                val users = snapshot?.documents?.mapNotNull { it.toUser() } ?: emptyList()
                trySend(users)
            }
        awaitClose { listener.remove() }
    }

    suspend fun updateFcmToken(userId: String, token: String) {
        firestore.collection(FirestoreConstants.USERS).document(userId)
            .update(FirestoreConstants.FCM_TOKEN, token).await()
    }
}

private fun com.google.firebase.firestore.DocumentSnapshot.toUser(): User? {
    val id = id
    val phone = getString(FirestoreConstants.PHONE) ?: return null
    val role = UserRole.fromString(getString(FirestoreConstants.ROLE)) ?: return null
    return User(
        id = id,
        phone = phone,
        role = role,
        displayName = getString(FirestoreConstants.DISPLAY_NAME),
        doctorId = getString(FirestoreConstants.DOCTOR_ID),
        fcmToken = getString(FirestoreConstants.FCM_TOKEN)
    )
}
